source/bigint.c

```diff
@@ -961,3 +1051,370 @@ int aws_bigint_shift_left(struct aws_bigint *bigint, size_t shift_amount) {
+
+        uint64_t two_digit_dividend = (current_remainder << BASE_BITS) + current_dividend_digit;
+
+        uint32_t quotient_digit = (uint32_t)(two_digit_dividend / wide_divisor);
+        aws_array_list_set_at(&temp_quotient->digits, &quotient_digit, quotient_length - 1 - i);
+
+        current_remainder = two_digit_dividend % wide_divisor;
+    }
+
+    s_aws_bigint_trim_leading_zeros(temp_quotient);
+
+    aws_array_list_swap_contents(&temp_quotient->digits, &quotient->digits);
+    quotient->sign = 1;
+
+    uint32_t final_remainder = (uint32_t)current_remainder;
+    aws_array_list_clear(&remainder->digits);
+    aws_array_list_push_back(&remainder->digits, &final_remainder);
+    remainder->sign = 1;
+
+    aws_bigint_destroy(temp_quotient);
+
+    AWS_POSTCONDITION(aws_bigint_is_valid(quotient));
+    AWS_POSTCONDITION(aws_bigint_is_valid(remainder));
+    AWS_POSTCONDITION(aws_bigint_is_valid(dividend));
+
+    return AWS_OP_SUCCESS;
+}
+
+/*
+ * The general division algorithm requires the divisor to be at least half the base.  In our implementation we shift
+ * the operands left (x 2) until the divisor's high bit is set.
+ */
+static uint32_t s_compute_divisor_normalization_shift(const struct aws_bigint *bigint) {
+    AWS_PRECONDITION(aws_bigint_is_valid(bigint));
+
+    size_t digit_count = aws_array_list_length(&bigint->digits);
+    AWS_FATAL_ASSERT(digit_count > 0);
+
+    uint32_t high_digit = 0;
+    aws_array_list_get_at(&bigint->digits, &high_digit, digit_count - 1);
+    AWS_FATAL_ASSERT(high_digit > 0);
+
+    uint32_t shift_count = 0;
+    while ((high_digit & DIVISION_NORMALIZATION_BIT_MASK) == 0) {
+        high_digit <<= 1;
+        ++shift_count;
+    }
+
+    AWS_POSTCONDITION(aws_bigint_is_valid(bigint));
+
+    return shift_count;
+}
+
+/*
+ * General algorithm requirements/assumptions:
+ *   (1) lhs >= rhs
+ *   (2) lhs, rhs both positive
+ *   (3) rhs is at least two digits in length
+ *   (4) lhs, rhs have been normalized so that rhs's high order digit has the high bit set
+ *   (5) lhs is at least one digit longer than rhs
+ *
+ * Additional points:
+ *   (1) lhs may temporarily have a leading zero digit (to satisfy requirement 5)
+ *
+ * Side effects:
+ *   (1) lhs is destructively modified to the remainder and then swapped with the "remainder" parameter
+ *
+ * Implementation based on Algorithm D, steps D2 - D7, in section 4.3.1 of AoCP Vol 2.
+ * Steps D1, D8 are performed by the caller.
+ */
+static int s_aws_bigint_normalized_divide(
+    struct aws_bigint **quotient,
+    struct aws_bigint **remainder,
+    struct aws_bigint *lhs,
+    const struct aws_bigint *rhs) {
+
+    AWS_PRECONDITION(aws_bigint_is_valid(lhs));
+    AWS_PRECONDITION(aws_bigint_is_valid(rhs));
+
+    size_t lhs_digit_count = aws_array_list_length(&lhs->digits);
+    size_t rhs_digit_count = aws_array_list_length(&rhs->digits);
+    AWS_FATAL_ASSERT(lhs_digit_count > rhs_digit_count); /* padding by zero */
```

Comment:
Couldn't they also be equal, as long as the most significant digit is lhs > rhs?

Reply:
We pad with a leading zero to prevent that.